File: lib/json_api/integration.rb
module JsonApi::Integration
  extend JsonApi::Json
  
  TYPE_KEY = 'integration'
  DEFAULT_PAGE = 10
  MAX_PAGE = 25
    
  def self.build_json(obj, args={})
    json = {}
    # TODO: include devices
    
    json['id'] = obj.global_id
    json['name'] = obj.settings['name']
    json['custom_integration'] = !!obj.settings['custom_integration']
    json['webhook'] = !!obj.settings['button_webhook_url']
    json['render'] = !!obj.settings['board_render_url']
    

    if args[:permissions]
      json['permissions'] = obj.permissions_for(args[:permissions])
    end

    if obj.integration_key
      json['integration_key'] = obj.integration_key
    end
    if json['permissions'] && json['permissions']['edit']
      if obj.template
        json['integration_key'] = obj.integration_key
        json['template'] = true
        if obj.settings['user_parameters']
          params = []
          obj.settings['user_parameters'].each do |param|
            param['type'] ||= 'text'
            params << param.slice('name', 'label', 'default_value', 'type', 'hint')
          end
          json['user_parameters'] = params
        end
      else
        json['added'] = obj.created_at.iso8601
        json['template_key'] = obj.settings['template_key']
        if obj.settings['user_settings']
          settings = []
          obj.settings['user_settings'].each do |key, data|
            setting = nil
            if data['type'] == 'password'
              setting = {'name' => key, 'label' => data['label'], 'protected' => true}
            else
              setting = {'name' => key, 'label' => data['label'], 'value' => data['value']}
            end
            settings << setting if setting
          end
          json['user_settings'] = settings
        end
      end

      if obj.settings['custom_integration']
        device_token = obj.device.token
        if obj.created_at > 24.hours.ago && obj.device
          json['access_token'] = device_token
          json['token'] = obj.settings['token']
        end
        json['truncated_access_token'] = "...#{device_token[-5, 5]}"
        json['truncated_token'] = "...#{obj.settings['token'][-5, 5]}"
      end
    end
    
    ['icon_url', 'description'].each do |key|
      json[key] = obj.settings[key] if obj.settings[key]
    end

    json
  end
  
  def self.extra_includes(obj, json, args={})
    json['integration']['render_url'] = obj.settings['board_render_url']
    if args[:permissions]
      json['integration']['user_token'] = obj.user_token(args[:permissions])
    end
    json
  end
end

File: app/frontend/app/models/integration.js
import DS from 'ember-data';
import CoughDrop from '../app';
import persistence from '../utils/persistence';
import { computed } from '@ember/object';

CoughDrop.Integration = DS.Model.extend({
  name: DS.attr('string'),
  user_id: DS.attr('string'),
  custom_integration: DS.attr('boolean'),
  webhook: DS.attr('boolean'),
  render: DS.attr('boolean'),
  render_url: DS.attr('string'),
  icon_url: DS.attr('string'),
  uses: DS.attr('number'),
  template: DS.attr('boolean'),
  template_key: DS.attr('string'),
  user_parameters: DS.attr('raw'),
  user_settings: DS.attr('raw'),
  integration_key: DS.attr('string'),
  description: DS.attr('string'),
  user_token: DS.attr('string'),
  button_webhook_url: DS.attr('string'),
  button_webhook_local: DS.attr('boolean'),
  board_render_url: DS.attr('string'),
  insecure_button_webhook_url: computed('button_webhook_url', 'button_webhook_local', function() {
    var url = this.get('button_webhook_url');
    return url && url.match(/^http:/) && !this.get('button_webhook_local');
  }),
  insecure_board_render_url: computed('board_render_url', function() {
    var url = this.get('board_render_url');
    return url && url.match(/^http:/);
  }),
  access_token: DS.attr('string'),
  truncated_access_token: DS.attr('string'),
  displayable_access_token: computed('access_token', 'truncated_access_token', function() {
    return this.get('access_token') || this.get('truncated_access_token');
  }),
  has_multiple_actions: computed('webhook', 'render', function() {
    return !!(this.get('webhook') && this.get('render'));
  }),
  token: DS.attr('string'),
  truncated_token: DS.attr('string'),
  displayable_token: computed('token', 'truncated_token', function() {
    return this.get('token') || this.get('truncated_token');
  }),
});

export default CoughDrop.Integration;
